### PR TITLE
Warp up prettier-eslint worker with first available open document

### PR DESCRIPTION
**Pull request by netux**
- Add "onStartupFinished" as an activation event for the extension.
- When the extension starts, wait until there is a document open and use it as a base to warm up the worker thread.

### CODE REVIEW

File: src/extension.js [hunk 1]
```diff
@@ -71,24 +92,43 @@ function formatter(document) {
   }
 }
 
-const supportedLanguages = [
```
Comment (JounQin) [inline src/extension.js:74]: Unrelated changes?

File: src/extension.js [hunk 2]
```diff
@@ -13,6 +13,27 @@ import requireRelative from 'require-relative';
...
+  'mdx',
+  'scss',
+  'svelte',
+  'typescript',
+  'typescriptreact',
+  'vue',
+  'yaml',
+];
+
+let prettierEslintPath;
```
Comment (JounQin) [inline src/extension.js:36]:
```suggestion
```

File: src/extension.js [hunk 3]
```diff
@@ -71,24 +92,43 @@ function formatter(document) {
...
+      const handler = window.onDidChangeActiveTextEditor(({ document }) => {
+        handler.dispose();
+        if (supportedLanguages.includes(document.languageId)) {
+          resolve(document);
+        }
+      });
+    })
+  } else {
+    return Promise.resolve(window.activeTextEditor.document);
+  }
```
Comment (JounQin) [inline src/extension.js:113]:
```suggestion
  }

  return Promise.resolve(window.activeTextEditor.document);
```

File: src/extension.js (hunk 1 above)
Comment (netux) [reply src/extension.js:74]: Since this was now being used by a helper method, I thought it'd make sense to put it to move it at the top of the file instead, before the method declarations.

File: src/extension.js (hunk 2 above)
Comment (netux) [reply src/extension.js:36]: Leftover from some testing I did.

File: src/extension.js (hunk 1 above)
Comment (JounQin) [reply src/extension.js:74]: Which method declaration you mean? I didn't see any reference used before declaration, and that's totally OK at the same time. My point is that don't change unnecessary codes.

File: src/extension.js [hunk 4]
```diff
@@ -97,3 +136,13 @@ supportedLanguages.forEach((language) => {
...
   });
 });
+
+waitForActiveSupportedDocument()
+  .then((document) => warmUpWorker(document))
+  .then(() => {
+    outputChannel.appendLine('Worker has been warmed up');
+  })
+  .catch((error) => {
+      console.error('Could not warm up worker with first available document:', error);
```
Comment (JounQin) [inline src/extension.js:146]: I'm not familiar with VSCode API, but how is this different from `outputChannel`? Can we only use one kind of logs?

File: src/extension.js (hunk 1 above)
Comment (netux) [reply src/extension.js:74]: It is used in `waitForActiveSupportedDocument()`. I moved it for organization purposes, as this is how I like to organize my constants, as is how I feel most people would organize their constants as well.

Since we seem to be on a stalemate, maybe we can ask @idahogurl what they think?

Comment (JounQin) [reply src/extension.js:74]: The diff shows the original codes is right up to `waitForActiveSupportedDocument`.

File: src/extension.js (hunk 4 above)
Comment (netux) [reply src/extension.js:146]: An output channel is shown in the Ui under the "OUTPUT" section on the bottom panel. AFAIK its a simple text buffer, with no formatting options, so it is more of an information to the user rather than a place to log stacktraces.


The `console.error` is only shown to developers while debugging the extension. It is here because the output of rejected Promises aren't logged to the console (or at least they shouldn't according to the spec).

Comment (idahogurl) [reply src/extension.js:146]: @netux Let's stick to the `outputChannel` because that will help the user provide us with the error and stack trace in case of debugging.

File: src/extension.js (hunk 1 above)
Comment (idahogurl) [reply src/extension.js:74]: @netux I don't have a preference either way. I tend to agree with @JounQin about making the least amount of changes. So if you wouldn't mind putting them above their first reference that would be great.

**Review by idahogurl — CHANGES_REQUESTED** [review]
Looks really good. Thank you! I addressed the concerns @JounQin brought up. If you'd like you can add yourself as a contributor in the README by using `yarn contributors:add` and following the prompts.

**Comment by netux** [comment]
Sorry for the delay, it's been a busy few days. I have done the changes as requested, as well as fix `waitForActiveSupportedDocument()` so it actually resolves if the opened document has a language that's supported.

File: src/extension.js [hunk 5]
```diff
@@ -97,3 +135,12 @@ supportedLanguages.forEach((language) => {
...
   });
 });
+
+waitForActiveSupportedDocument()
+  .then((document) => warmUpWorker(document))
+  .then(() => {
+    outputChannel.appendLine('Worker has been warmed up');
+  })
+  .catch((error) => {
+      outputChannel.appendLine('Error: Could not warm up worker. Formatting a file for the first time may take longer than usual.\nStacktrace: ' + error.stack);
```
Comment (JounQin) [inline src/extension.js:145]: Will `error.message` be available in `error.stack`?

**Review by JounQin — APPROVED** [review]
